File: leetcode/138-copy-list-with-random-pointer.py
# https://leetcode.com/problems/copy-list-with-random-pointer/ (top-interview-questions)


from typing import Optional

# Definition for a Node.
class Node:
	def __init__(self, x: int, next = None, random = None):
		self.val = int(x)
		self.next = next
		self.random = random

class Solution:
	def copyRandomList(self, head: Optional[Node]) -> Optional[Node]:
		nodes_map = { None: None } # original : copy

		curr = head
		while curr:
			nodes_map[curr] = Node(curr.val)
			curr = curr.next

		curr = head
		while curr:
			nodes_map[curr].next = nodes_map[curr.next]
			nodes_map[curr].random = nodes_map[curr.random]
			curr = curr.next

		return nodes_map[head]

File: leetcode/138-copy-list-with-random-pointer.js
// https://leetcode.com/problems/copy-list-with-random-pointer/ (top-interview-questions)

// Definition for a Node.
function Node(val, next, random) {
  this.val = val;
  this.next = next;
  this.random = random;
};

/**
 * @param {Node} head
 * @return {Node}
 */
var copyRandomList = function(head) {

  // create a mapping of original node to new node
  let map = new Map();
  let current = head;
  while (current) {
      map.set(current, new Node(current.val, null, null));
      current = current.next;
  }
  
  // set next and random pointers of new nodes
  current = head;
  while (current) {
      map.get(current).next = map.get(current.next);
      map.get(current).random = map.get(current.random);
      current = current.next;
  }
  
  return map.get(head);

};
